### PR TITLE
EVG-12438: Implement gRPC CreateSystemMetrics and CloseSystemMetrics

### PR DESCRIPTION
**Pull request by maravichandran**

### CODE REVIEW

File: model/system_metrics.go [hunk 1]
```diff
@@ -244,6 +244,44 @@ func (sm *SystemMetrics) appendSystemMetricsChunkKey(ctx context.Context, key st
 	return errors.Wrapf(err, "problem appending system metrics data chunk to %s", sm.ID)
 }
 
+// Close "closes out" the log by populating the completed_at and info.exit_code
```
Comment (Kimchelly) [inline model/system_metrics.go:247]: Remove `info.exit_code`.

File: model/system_metrics_test.go [hunk 2]
```diff
@@ -396,3 +396,62 @@ func TestSystemMetricsSaveNew(t *testing.T) {
...
+		e1 := 0
+		l1 := &SystemMetrics{ID: sm1.ID, populated: true}
+		l1.Setup(env)
+		require.NoError(t, l1.Close(ctx, e1))
+
+		updatedSystemMetrics := &SystemMetrics{}
+		require.NoError(t, db.Collection(buildloggerCollection).FindOne(ctx, bson.M{"_id": sm1.ID}).Decode(updatedSystemMetrics))
+		assert.Equal(t, sm1.ID, updatedSystemMetrics.ID)
+		assert.Equal(t, sm1.ID, updatedSystemMetrics.Info.ID())
+		assert.Equal(t, sm1.CreatedAt.UTC().Round(time.Second), updatedSystemMetrics.CreatedAt.Round(time.Second))
```
Comment (Kimchelly) [inline model/system_metrics_test.go:435]: Can this just be `assert.WithinDuration`?

File: model/system_metrics_test.go [hunk 3]
```diff
@@ -396,3 +396,62 @@ func TestSystemMetricsSaveNew(t *testing.T) {
...
+		e2 := 9
+		l2 := &SystemMetrics{Info: sm2.Info, populated: true}
+		l2.Setup(env)
+		require.NoError(t, l2.Close(ctx, e2))
+
+		updatedSystemMetrics := &SystemMetrics{}
+		require.NoError(t, db.Collection(buildloggerCollection).FindOne(ctx, bson.M{"_id": sm2.ID}).Decode(updatedSystemMetrics))
+		assert.Equal(t, sm2.ID, updatedSystemMetrics.ID)
+		assert.Equal(t, sm2.ID, updatedSystemMetrics.Info.ID())
+		assert.Equal(t, sm2.CreatedAt.UTC().Round(time.Second), updatedSystemMetrics.CreatedAt.Round(time.Second))
```
Comment (Kimchelly) [inline model/system_metrics_test.go:451]: Same question as above.

File: rpc/internal/formats.ext.go [hunk 4]
```diff
@@ -0,0 +1,50 @@
+package internal
+
+import "github.com/evergreen-ci/cedar/model"
+
+func (f DataFormat) Export() model.FileDataFormat {
```
Comment (Kimchelly) [inline rpc/internal/formats.ext.go:5]: Can we remove this or is the assumption that it'll be removed in a different PR?

File: rpc/internal/system_metrics.ext.go [hunk 5]
```diff
@@ -0,0 +1,29 @@
...
+		Execution: int(sm.Execution),
+		Mainline:  sm.Mainline,
+	}
+}
+
+// Export exports SystemMetricsArtifactInfo to the corresponding SystemMetricsInfo
+// type in the model package.
+func (sm SystemMetricsArtifactInfo) Export() model.SystemMetricsArtifactOptions {
+	return model.SystemMetricsArtifactOptions{
+		Format:      sm.Format.Export(),
```
Comment (Kimchelly) [inline rpc/internal/system_metrics.ext.go:25]: Same question about format.

File: rpc/internal/system_metrics_service.go [hunk 6]
```diff
@@ -26,9 +26,20 @@ func AttachSystemMetricsService(env cedar.Environment, s *grpc.Server) {
...
-//
-func (s *systemMetricsService) CreateSystemMetricRecord(context.Context, *SystemMetrics) (*SystemMetricsResponse, error) {
-	return nil, nil
+// CreateSystemMetricRecord creates a new system metrics record in the database.
+func (s *systemMetricsService) CreateSystemMetricRecord(ctx context.Context, data *SystemMetrics) (*SystemMetricsResponse, error) {
+	conf := model.NewCedarConfig(s.env)
+	if err := conf.Find(); err != nil {
+		return nil, newRPCError(codes.Internal, errors.Wrap(err, "problem fetching cedar config"))
+	}
+	if conf.Bucket.SystemMetricsBucketType == "" {
```
Comment (Kimchelly) [inline rpc/internal/system_metrics_service.go:35]: I know this is probably copy-pasted from test results, but is it actually a reasonable assumption for the default bucket type to be a local filesystem bucket?

File: rpc/internal/system_metrics_service.go [hunk 7]
```diff
@@ -78,7 +89,19 @@ func (s *systemMetricsService) StreamSystemMetrics(stream CedarSystemMetrics_Str
 	}
 }
 
-//
-func (*systemMetricsService) CloseMetrics(context.Context, *SystemMetricsSeriesEnd) (*SystemMetricsResponse, error) {
-	return nil, nil
+// CloseMetrics "closes out" a system metrics record by setting the
+// completed at timestamp. This should be the last rcp call made on a system
```
Comment (Kimchelly) [inline rpc/internal/system_metrics_service.go:93]: rpc*

File: rpc/internal/formats.ext.go (hunk 4 above)
Comment (julianedwards) [reply rpc/internal/formats.ext.go:5]: We still need it for perf

File: rpc/internal/system_metrics_service.go (hunk 6 above)
Comment (julianedwards) [reply rpc/internal/system_metrics_service.go:35]: Yeah for testing

Comment (Kimchelly) [reply rpc/internal/system_metrics_service.go:35]: Sure, but is it actually reasonable to mix in test code with production code? Shouldn't the tests/local cedar specify this in the cedar configuration?

File: rpc/internal/system_metrics_service.go (hunk 7 above)
Comment (maravichandran) [reply rpc/internal/system_metrics_service.go:93]: done

File: model/system_metrics.go (hunk 1 above)
Comment (maravichandran) [reply model/system_metrics.go:247]: done

File: model/system_metrics_test.go (hunk 2 above)
Comment (maravichandran) [reply model/system_metrics_test.go:435]: done

File: model/system_metrics_test.go (hunk 3 above)
Comment (maravichandran) [reply model/system_metrics_test.go:451]: done

File: rpc/internal/system_metrics_service.go (hunk 6 above)
Comment (maravichandran) [reply rpc/internal/system_metrics_service.go:35]: done

File: model/system_metrics_test.go [hunk 8]
```diff
@@ -396,3 +396,61 @@ func TestSystemMetricsSaveNew(t *testing.T) {
...
+		l1 := &SystemMetrics{ID: sm1.ID, populated: true}
+		l1.Setup(env)
+		require.NoError(t, l1.Close(ctx))
+
+		updatedSystemMetrics := &SystemMetrics{}
+		require.NoError(t, db.Collection(systemMetricsCollection).FindOne(ctx, bson.M{"_id": sm1.ID}).Decode(updatedSystemMetrics))
+		assert.Equal(t, sm1.ID, updatedSystemMetrics.ID)
+		assert.Equal(t, sm1.ID, updatedSystemMetrics.Info.ID())
+		assert.WithinDuration(t, sm1.CreatedAt.UTC(), updatedSystemMetrics.CreatedAt, time.Second)
+		//assert.Equal(t, sm1.CreatedAt.UTC().Round(time.Second), updatedSystemMetrics.CreatedAt.Round(time.Second))
```
Comment (Kimchelly) [inline model/system_metrics_test.go:435]: Remove line.

File: rpc/internal/system_metrics.ext.go (hunk 5 above)
Comment (Kimchelly) [reply rpc/internal/system_metrics.ext.go:25]: This should still be removed since we're no longer using format.

See [this](https://github.com/evergreen-ci/timber/pull/30/files#diff-5921bc410dd2932ed31eef8df216a63eL27)

File: model/system_metrics_test.go (hunk 8 above)
Comment (maravichandran) [reply model/system_metrics_test.go:435]: done

File: rpc/internal/system_metrics.ext.go (hunk 5 above)
Comment (maravichandran) [reply rpc/internal/system_metrics.ext.go:25]: done
